src/main/java/org/veupathdb/service/eda/ss/model/reducer/FormattedTabularRecordStreamer.java

```diff
@@ -0,0 +1,124 @@
+package org.veupathdb.service.eda.ss.model.reducer;
+
+import org.veupathdb.service.eda.ss.model.Entity;
+import org.veupathdb.service.eda.ss.model.variable.VariableValueIdPair;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Provides a stream of variable value records extracted from a stream of ID indexes and a collection of
+ * ID index, variable value pairs. The resulting stream provides formatted tuples of strings containing entity IDs,
+ * ancestor IDs and all associated variable values.
+ */
+public class FormattedTabularRecordStreamer implements Iterator<List<String>> {
+  private List<ValueStream> valuePairStreams;
+  private ValueStream<List<Long>> ancestorStream;
+  private Iterator<Long> idIndexStream;
+  private Long currentIdIndex;
+  private Entity entity;
+
+  /**
+   * Constructs an instance, which provides a stream of string-formatted records. The stream is composed of all IDs
+   * in the idIndex stream, corresponding ancestor IDs corresponding values in the
+   * {@link FormattedTabularRecordStreamer#valuePairStreams}
+   *
+   * @param valuePairStreams Streams of pairs containing entity ID indexes and corresponding values for their respective
+   *                         variables.
+   * @param idIndexStream    Stream of ID indexes, indicating which entity records to output.
+   * @param ancestorStream   Stream of ancestors, used to merge in ancestors of target output entities in output records.
+   * @param entity           Entity type of records which will be output.
+   */
+  public FormattedTabularRecordStreamer(List<Iterator<VariableValueIdPair<?>>> valuePairStreams,
+                                        Iterator<Long> idIndexStream,
+                                        Iterator<VariableValueIdPair<List<Long>>> ancestorStream,
+                                        Entity entity) {
+    this.valuePairStreams = valuePairStreams.stream()
+        .map(s -> new ValueStream(s))
+        .collect(Collectors.toList());
+    this.idIndexStream = idIndexStream;
+    if (idIndexStream.hasNext()) {
+      currentIdIndex = idIndexStream.next();
+    }
+    this.ancestorStream = new ValueStream<>(ancestorStream);
+    this.entity = entity;
+  }
+
+  @Override
+  public boolean hasNext() {
+    return currentIdIndex != null;
+  }
+
+  @Override
+  public List<String> next() {
+    List<String> record = new ArrayList<>();
+    // Only add ancestors if entity has ancestors associated with it.
+    if (!entity.getAncestorEntities().isEmpty()) {
+      if (!ancestorStream.hasNext()) {
+        throw new NoSuchElementException("No tabular records remain in stream.");
+      }
+      while (ancestorStream.peek().getIdIndex() < currentIdIndex) {
+        ancestorStream.next();
+      }
+      // Add all ancestor IDs
```

Comment:
need to check the order of id and ancestors.   is it:  id, parent, grandparent, greatgp   or greatgp, gp, parent, id?

the binary file dumper seems to put the id at the left